onnx/defs/parser.h

```diff
@@ -0,0 +1,382 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+// Experimental language syntax and parser for ONNX. Please note that the syntax as formalized
+// by this parser is preliminary and may change.
+
+#pragma once
+
+#include <ctype.h>
+#include <iostream>
+#include <stdexcept>
+#include <string>
+#include <unordered_map>
+
+#include "onnx/onnx_pb.h"
+
+#include "onnx/common/status.h"
+#include "onnx/string_utils.h"
+
+namespace ONNX_NAMESPACE {
+
+using namespace ONNX_NAMESPACE::Common;
+
+using IdList = google::protobuf::RepeatedPtrField<std::string>;
+
+using NodeList = google::protobuf::RepeatedPtrField<NodeProto>;
+
+using AttrList = google::protobuf::RepeatedPtrField<AttributeProto>;
+
+using ValueInfoList = google::protobuf::RepeatedPtrField<ValueInfoProto>;
+
+#define PARSE_ERROR(...)                                                                                        \
+  do {                                                                                                          \
+    return Status(                                                                                              \
+        NONE, FAIL, ONNX_NAMESPACE::MakeString("[ParseError at position ", GetCurrentPos(), "]", __VA_ARGS__)); \
+  } while (0)
```

Comment:
This macro looks kina hacky. You can achieve the same result by using a variadic template or some other construct that will be type safe.

Reply:
Are you referring to MakeString? I am reusing what's being used elsewhere in ONNX. The outer macro was just to hide whether exceptions or status is being used to handle errors. The original implementation did use exceptions, but was switched to use Status due to the overall ONNX goals.

Reply:
No, I meant the whole outer macro. Inside it has this construct with do..while(0), which I've seen many years ago and it was used to adress some specific problem but in a really hacky way. Can this just become a variadic template that builds an error message?

Reply:
Done.